stan/math/rev/fun/trace_quad_form.hpp

```diff
@@ -93,6 +95,94 @@ inline return_type_t<EigMat1, EigMat2> trace_quad_form(const EigMat1& A,
       new internal::trace_quad_form_vari<Ta, Ra, Ca, Tb, Rb, Cb>(baseVari));
 }
 
+/**
+ * Compute trace(B^T A B).
+ *
+ * This overload handles arguments where one of Mat1 or Mat2 are
+ * `var_value<T>` where `T` is an Eigen type. The other type can
+ * also be a `var_value` or it can be a type that inherits
+ * from EigenBase
+ *
+ * @tparam Mat1 type of the first matrix
+ * @tparam Mat2 type of the second matrix
+ *
+ * @param A matrix
+ * @param B matrix
+ * @return The trace of B^T A B
+ * @throw std::domain_error if A is not square
+ * @throw std::domain_error if A cannot be multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline var trace_quad_form(const Mat1& A, const Mat2& B) {
+  check_square("trace_quad_form", "A", A);
+  check_multiplicable("trace_quad_form", "A", A, "B", B);
+
+  var res;
+
+  if (!is_constant<Mat1>::value && !is_constant<Mat2>::value) {
+    arena_t<promote_scalar_t<var, Mat1>> arena_A = A;
+    arena_t<promote_scalar_t<var, Mat2>> arena_B = B;
+
+    res = (value_of(arena_B).transpose() * value_of(arena_A)
+           * value_of(arena_B))
+              .trace();
+
+    reverse_pass_callback([arena_A, arena_B, res]() mutable {
+      if (is_var_matrix<Mat1>::value) {
+        arena_A.adj().noalias()
+            += res.adj() * value_of(arena_B) * value_of(arena_B).transpose();
+      } else {
+        arena_A.adj()
+            += res.adj() * value_of(arena_B) * value_of(arena_B).transpose();
+      }
```

Comment:
[side note] I wonder if it would be useful or look nice to have something like

```cpp
update_adjoint(arena_A, res.adj() * value_of(arena_B) * value_of(arena_B).transpose());
```

Where for `var<mat>` it's just

```cpp
lhs.adj().nolias() += rhs
```

and for `mat<var>` it is 

```cpp
lhs.adj() += rhs
```

So we don't have to have these if-else statements all over

Reply:
What about:

`noalias(lhs.adj()) += rhs`

Reply:
Yeah what I said doesn't make sense cause you'd still need the `is_var_matrix` logic.

I can add this if you want. I'm happy with ifs though.